### PR TITLE
feat: token order input interactivity

### PR DESCRIPTION
**Pull request by gambinish**
## **Description**

Token side of the order input was non-interactive, meaning that users could only input orders as dollar amounts. This PR introduces that interactivity so that users can enter either/both amounts. Updating one input will correspond with a "connected" update with the other.

## **Changelog**

CHANGELOG entry: Make order input for token amount interactive

## **Related issues**

Fixes: https://consensyssoftware.atlassian.net/browse/TAT-3002

## **Manual testing steps**

Enter an order with either: dollar amount, token amount, or slider. All elements should update in unison.

## **Screenshots/Recordings**

https://github.com/user-attachments/assets/31974490-d225-4b4e-ad24-90098f941b07

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

---

> [!NOTE]
> **Medium Risk**
> Touches order sizing input logic and formatting, which can affect calculated USD size/percent values; changes are localized and covered by new tests but could impact edge-case entry behavior across locales.
> 
> **Overview**
> Enables fully interactive token-side sizing in perps order entry by making the token input behave like a real editable field (preserving intermediate typing states like `0`/`0.`), while keeping USD size and percent slider calculations in sync.
> 
> Introduces `formatNumberForInput` to render token values in a locale-neutral, non-grouped `.`-decimal format so validation/backspace editing works for large amounts, and adds focused unit tests covering partial input preservation, clearing behavior, large-number formatting, and external amount-to-token resync when not editing.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 3ffda826d25c2194654d3b5e5b7b76f4ab99228c. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: ui/components/app/perps/order-entry/components/amount-input/amount-input.tsx [hunk 1]
```diff
@@ -70,16 +76,31 @@ export const AmountInput: React.FC<AmountInputProps> = ({
...
   }, [amount, currentPrice]);
 
-  const tokenDisplayValue = useMemo(() => {
+  // Un-grouped so the string is always editable by isUnsignedDecimalInput (no commas).
+  const unGroupedTokenDisplay = useMemo(() => {
     if (tokenAmount === null || tokenAmount === 0) {
       return '';
     }
     return formatNumber(tokenAmount, {
+      useGrouping: false,
```
[the remark this answers is not in the document]
Comment (gambinish) [reply ui/components/app/perps/order-entry/components/amount-input/amount-input.tsx]: https://github.com/MetaMask/metamask-extension/pull/42013/commits/52623014449ceb4d3c5f2dfe6cf1fd4e8fa75600

File: ui/components/app/perps/order-entry/components/amount-input/amount-input.tsx [hunk 2]
```diff
@@ -70,15 +80,27 @@ export const AmountInput: React.FC<AmountInputProps> = ({
...
+  }, [tokenAmount]);
+
+  // Local draft for the token input so intermediate values (e.g. "0", "0.") are
+  // preserved while the user is actively typing.
+  const [tokenInputValue, setTokenInputValue] = useState(unGroupedTokenDisplay);
+  const isEditingTokenRef = useRef(false);
+
+  // Keep the token input in sync with external state changes (slider, USD field,
+  // percent pill) unless the user is currently typing in the token field.
+  useEffect(() => {
```
Comment (geositta) [inline ui/components/app/perps/order-entry/components/amount-input/amount-input.tsx]: This is the same React docs case AGENTS.md warns about: using useEffect to sync derived state. Right now, when the slider or USD field updates amount, the component first renders with stale tokenInputValue, then the effect schedules a second render with the synced value.

Instead of patching that, prefer we remove it. Derive the displayed token value during render, and keep local state only for the active edit draft:

```
const [isEditingToken, setIsEditingToken] = useState(false);
const [tokenInputValue, setTokenInputValue] =
  useState(unGroupedTokenDisplay);

const displayedTokenValue = isEditingToken
  ? tokenInputValue
  : unGroupedTokenDisplay;

const handleTokenFocus = useCallback(() => {
  setTokenInputValue(unGroupedTokenDisplay);
  setIsEditingToken(true);
}, [unGroupedTokenDisplay]);

const handleTokenBlur = useCallback(() => {
  setIsEditingToken(false);
}, []);
```

The focus handler still needs to seed the draft from unGroupedTokenDisplay.

Otherwise, if the user typed something earlier, blurred the field, and then amount changed from some other control, focusing the token input again can bring back that stale draft instead of starting from the latest derived value. 

With the value chosen in render, blur becomes simpler: it only needs to mark editing as finished, because the nonediting path will show the current derived display.

I think that would leave the component in a better place overall. It removes the effect driven sync path, avoids the extra render that comes with it, and makes the state model easier to reason about: one piece of state tracks whether the user is actively editing, one holds the temporary draft, and the derived display stays derived.

**Review by geositta — COMMENTED** [review]
Would approve but I would like to see the useEffect removed first.

File: ui/components/app/perps/order-entry/components/amount-input/amount-input.tsx (hunk 2 above)
Comment (gambinish) [reply ui/components/app/perps/order-entry/components/amount-input/amount-input.tsx]: https://github.com/MetaMask/metamask-extension/pull/42013/commits/240a455e4bf418dcdbb6ce0e302ad261eabfd61e

**Review by aganglada — APPROVED** [review]

**Review by geositta — APPROVED** [review]
Thanks for refactoring the useEffect, looks good.
